src/Conversion/ONNXToKrnl/Tensor/Shape.cpp

```diff
@@ -0,0 +1,67 @@
+//===----------------Shape.cpp - Lowering Shape Op----------------------=== //
+//
+// Copyright 2020 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file lowers the ONNX Shape Operator to Krnl dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp"
+
+using namespace mlir;
+
+struct ONNXShapeOpLowering : public ConversionPattern {
+  ONNXShapeOpLowering(MLIRContext *ctx)
+      : ConversionPattern(mlir::ONNXShapeOp::getOperationName(), 1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    ONNXShapeOpAdaptor operandAdaptor(operands);
+    ONNXShapeOp shapeOp = llvm::cast<ONNXShapeOp>(op);
+    bool insertDealloc = checkInsertDealloc(op);
+    Location loc = op->getLoc();
+    // Get input data.
+    Value data = operandAdaptor.data();
+    ArrayRef<int64_t> dataShape = data.getType().cast<MemRefType>().getShape();
+    unsigned dataRank = dataShape.size();
+
+    Value resultOperand = shapeOp.shape();
+    MemRefType outputMemRefType = convertToMemRefType(*op->result_type_begin());
+
+    Value alloc;
+    if (hasAllConstantDimensions(outputMemRefType))
+      alloc =
+          insertAllocAndDealloc(outputMemRefType, loc, rewriter, insertDealloc);
+    else
+      alloc = insertAllocAndDealloc(
+          outputMemRefType, loc, rewriter, insertDealloc, {resultOperand});
```

Comment:
I don't understand why resultOperand is in tensorType. At this level of abstraction, wouldn't all tensor values be lowered to memref values? @tungld @doru1004 may have more insight.

Reply:
With the current design, tensor-to-memref conversion is combined with the onnx-to-krnl pass. As a result, only input tensors are converted to memref values, and the output (in tensorType) is replaced by `alloc` in memref to be consistent with the next operation's inputs. Since `shapeOp.shape` is the output, so it is in tensorType.

As I know, @doru1004 has a tensor-to-memref pass. We can call it before the onnx-to-krnl pass, then all tensor values at this level will be in memref type.